homeassistant/components/somfy/config_flow.py

```diff
@@ -0,0 +1,146 @@
+"""Config flow for Somfy."""
+import asyncio
+import logging
+
+import async_timeout
+
+from homeassistant import config_entries
+from homeassistant.components.http import HomeAssistantView
+from homeassistant.core import callback
+from .const import CLIENT_ID, CLIENT_SECRET, DOMAIN
+
+AUTH_CALLBACK_PATH = '/auth/somfy/callback'
+AUTH_CALLBACK_NAME = 'auth:somfy:callback'
+
+_LOGGER = logging.getLogger(__name__)
+
+
+@callback
+def register_flow_implementation(hass, client_id, client_secret):
+    """Register a flow implementation.
+    client_id: Client id.
+    client_secret: Client secret.
+    """
+    hass.data[DOMAIN][CLIENT_ID] = client_id
+    hass.data[DOMAIN][CLIENT_SECRET] = client_secret
+
+
+@config_entries.HANDLERS.register('somfy')
+class SomfyFlowHandler(config_entries.ConfigFlow):
+    """Handle a config flow."""
+
+    VERSION = 1
+    CONNECTION_CLASS = config_entries.CONN_CLASS_CLOUD_POLL
+
+    async def async_step_import(self, user_input=None):
+        """Handle external yaml configuration."""
+        if self.hass.config_entries.async_entries(DOMAIN):
+            return self.async_abort(reason='already_setup')
+        return await self.async_step_auth()
+
+    async def async_step_user(self, user_input=None):
+        """Handle a flow start."""
+        return await self.async_step_auth()
+
+    async def async_step_auth(self, user_input=None):
+        """Create an entry for auth."""
+        if DOMAIN not in self.hass.data:
+            return self.async_abort(reason='no_flows')
+
+        if self.hass.config_entries.async_entries(
+                DOMAIN) and self.cur_step is None:
```

Comment:
When will this happen?

Does this guard from user set up with already configured entry?

Reply:
Yes that the case. When the end user click on Configure, and we have already a Somfy config entry, self.cur_step is None and a nice message explain we can have only one instance.

Reply:
If you have a better way to do this, I can change it.

Reply:
When would we have a config entry and we shouldn't abort? Why do we need the `cur_step` check?

Reply:
I agree we should abort as soon as we have a config entry. And it's the case here. The line after, I only check if there is already a config entry. There is two cases in order to have two different messages. 

**Scenario**:
* End user click for the first time on Configure
* A modal window invite it to click on a link to authorize HASS on Somfy side. The link is opened 
within a new tab.
* Somfy redirects to HASS instance.
* This is caught by *SomfyAuthCallbackView* (redirecting to the HASS home page). This last one init a flow at step *code*.
* Step *code*, retrieve the code parameter from the URL, ask Somfy for a token and create the config entry. 
* End user can close its tab.
* Back on the first tab, there is still the modal window. End user click on Submit.

When the end user click on the Submit, I don't want to show it the message *You can configure only one instance*. It sound like an error in this context. Instead *An external flow already configured Somfy.* make more sense.

It's why there is two different tests.

The best solution would be to have a method to continue the flow using a View.

Reply:
We should try to keep track of the ongoing flow and close it automatically when the entry is created.

Reply:
The point component flow handles this by having to different reasons for aborting with already created entry. But it checks that early in user step, and aborts with already created. Here in auth step we just need to abort with external setup if there is an entry.

Reply:
You are totally right. I've moved this code to async_step_user. It's more elegant.

Reply:
So I have a WIP branch that allows continuing an existing flow. It works by having a new result type "External Step". The external step would end up triggering something inside HA (like your webview) and then we could tell config entry to proceed, notifying the frontend via an event.

Code is here https://github.com/home-assistant/home-assistant/commit/5efbe119f974d3e12a8ce697a456a9a1bcd18f52#diff-57c19b8c04f9d3fbed49f0e00508d04fR194

Reply:
The way it works is that you store the config flow ID inside the OAuth state. The OAuth server will return this state when it hits our callback view. Then we use that to pass the data to the config flow.

```
        await hass.config_entries.flow.async_configure(
            payload['config_flow_id'], payload['tokens'])
```

Reply:
Feel free to copy the relevant changes of that branch into this PR to make it work.

Reply:
Seems great! I will study the piece of code you gave me and come back later if something is unclear.

Reply:
OK, so I've updated my view to retrieve the flow id from the state. 

```python
class SomfyAuthCallbackView(HomeAssistantView):
    """Somfy Authorization Callback View."""

    requires_auth = False
    url = AUTH_CALLBACK_PATH
    name = AUTH_CALLBACK_NAME

    @staticmethod
    async def get(request):
        """Receive authorization code."""
        from aiohttp import web
        hass = request.app['hass']
        if 'code' in request.query:
           # continue flow, go back to auth step
            hass.async_create_task(
                hass.config_entries.flow.async_configure(
                    request.query['state'], {'code': request.query['code']})

            )
        return web.HTTPFound('/')
```

But I don't understand how I have to use your new external step. What's the purpose of the url key? Whatever the value I tried, in the front end I just get an blank modal. I thought it was used to redirect the end user to an external page. So I guess I don't return the external step at the correct place.

```python
    async def async_step_auth(self, user_input=None):
        """Create an entry for auth."""
        if self.hass.config_entries.async_entries(DOMAIN):
            return self.async_abort(reason='external_setup')

        if user_input is not None and 'code' in user_input:
            return await self.async_step_code(user_input['code'])

        try:
            with async_timeout.timeout(10):
                url, _ = await self._get_authorization_url(self.flow_id)
        except asyncio.TimeoutError:
            return self.async_abort(reason='authorize_url_timeout')
        except Exception:  # pylint: disable=broad-except
            _LOGGER.exception("Unexpected error generating auth url")
            return self.async_abort(reason='authorize_url_fail')

        return self.async_external_step(
            step_id='auth',
            url=url,
        )
```

where url look like `https://accounts.somfy.com/oauth/oauth/v2/auth?response_type=code&client_id=xxxx&redirect_uri=http%3A%2F%2F192.168.1.87%3A8123%2Fauth%2Fsomfy%2Fcallback&state=ff14af95d8ac4e9e9c867c5aa92ec6d2`

Do you expect the end user to still click on a link?

Reply:
@balloob

Reply:
@MartinHjelmare  @balloob Can this new step work without any modification of the UI?

Reply:
> Whatever the value I tried, in the front end I just get an blank modal. I thought it was used to redirect the end user to an external page.

I had the same issue with the Point implementation and never managed to fix that, the page shown at the redirect url is just an empty page. BUT this in my view:

```python
hass.async_create_task(
                hass.config_entries.flow.async_init(
                    DOMAIN,
                    context={'source': 'code'},
                    data=request.query['code'],
                ))
```

Calls `async def async_step_code(self, code=None):` with code as the code I got after the external auth, so that my configuration can continue at https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/point/config_flow.py#L125

Reply:
@fredrike Thx for your comment and sorry for the delay (i missed the notification). Your solution was exactly what I tried in a first time. I think I will have to do it again.

Reply:
This is how I document the flow in [Point](https://www.home-assistant.io/components/point), it is not optimal but it works fine until we have support for cloud code auth.

Once Home Assistant is started, a configurator will pop up asking you to Authenticate your Point account via a link. When you follow the link and click on **Accept** you will be redirected to the `callback url` and the Point integration will be automatically configured and you can go back to the original dialog and press **Submit**.

Reply:
That's exactly the behavior I have thx. I will document it too. As you are here, can you explain why your view return [OK](https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/point/config_flow.py#L189). When I do this, and when Somfy redirect me to the callback, my browser ask me to download a file... So for the moment, I redirect to the HASS home page.